File: scripts/import-methodology.py
#!/usr/bin/env python3
"""Import Ars Contexta methodology (249 research claims) into a Knowledge Vault via Switchboard CLI."""

import hashlib
import json
import os
import re
import subprocess
import sys
import time
METHODOLOGY_DIR = os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))), "data", "methodology")
SW = os.environ.get("SWITCHBOARD_BIN", "switchboard")


def run_sw(*args, timeout=60, retries=2):
    """Run a switchboard CLI command and return parsed JSON or raw output."""
    cmd = [SW] + list(args)
    for attempt in range(retries + 1):
        try:
            result = subprocess.run(cmd, capture_output=True, text=True, timeout=timeout)
            if result.returncode != 0:
                if attempt < retries and ("timeout" in result.stderr.lower() or "connection" in result.stderr.lower()):
                    time.sleep(2 * (attempt + 1))
                    continue
                return None, result.stderr
            try:
                return json.loads(result.stdout), None
            except json.JSONDecodeError:
                return result.stdout.strip(), None
        except subprocess.TimeoutExpired:
            if attempt < retries:
                time.sleep(2 * (attempt + 1))
                continue
            return None, "Command timed out"
    return None, "All retries exhausted"


def run_sw_raw(*args, timeout=30):
    """Run a switchboard CLI command and return raw stdout."""
    cmd = [SW] + list(args)
    result = subprocess.run(cmd, capture_output=True, text=True, timeout=timeout)
    return result.stdout, result.stderr, result.returncode


def parse_methodology_file(filepath):
    """Parse a methodology markdown file into structured data."""
    with open(filepath, "r") as f:
        content = f.read()

    # Split YAML frontmatter from body (no pyyaml dependency)
    parts = content.split("---")
    if len(parts) < 3:
        return None

    fm_text = parts[1]
    frontmatter = {}
    for line in fm_text.strip().split("\n"):
        m = re.match(r'^(\w+):\s*(.+)$', line)
        if m:
            key, val = m.group(1), m.group(2).strip()
            # Parse arrays like ["a", "b"] or [a, b]
            if val.startswith("["):
                items = re.findall(r'"([^"]+)"|\'([^\']+)\'|\[\[([^\]]+)\]\]|(\w[\w\s-]+)', val)
                frontmatter[key] = [next(g for g in groups if g) for groups in items]
            elif val.startswith('"') and val.endswith('"'):
                frontmatter[key] = val.strip('"')
            elif val.startswith("'") and val.endswith("'"):
                frontmatter[key] = val.strip("'")
            elif val.startswith("[[") and val.endswith("]]"):
                frontmatter[key] = val[2:-2]
            else:
                frontmatter[key] = val
    body = "---".join(parts[2:]).strip()

    # Extract title from filename
    title = os.path.basename(filepath).replace(".md", "")

    # Split body into content and relevant notes
    body_content = body
    connections = []

    # Find "Relevant Notes:" section
    rn_match = re.search(r"\n(?:Relevant Notes|---)\s*\n", body)
    if rn_match:
        body_content = body[: rn_match.start()].strip()
        notes_section = body[rn_match.end() :]

        # Parse connection lines: - [[target]] — context
        for line in notes_section.split("\n"):
            m = re.match(
                r"^-\s+\[\[(.+?)\]\]\s*[—–-]\s*(.+)$", line.strip()
            )
            if m:
                connections.append(
                    {"target": m.group(1).strip(), "context": m.group(2).strip()}
                )

    # Remove the # Title line from content if it matches the filename
    lines = body_content.split("\n")
    if lines and lines[0].startswith("# "):
        body_content = "\n".join(lines[1:]).strip()

    # Extract topics from frontmatter
    topics = []
    if frontmatter.get("topics"):
        for t in frontmatter["topics"]:
            # Remove [[]] wrappers
            clean = re.sub(r"\[\[|\]\]", "", str(t)).strip()
            if clean:
                topics.append(clean)

    # Extract methodology
    methodology = frontmatter.get("methodology", [])
    if isinstance(methodology, str):
        methodology = [methodology]

    # Extract sources
    sources = []
    src = frontmatter.get("source", "")
    if src:
        clean_src = re.sub(r"\[\[|\]\]", "", str(src)).strip()
        if clean_src:
            sources.append(clean_src)

    return {
        "title": title,
        "description": frontmatter.get("description", ""),
        "content": body_content,
        "kind": frontmatter.get("kind", "research"),
        "methodology": methodology,
        "sources": sources,
        "topics": topics,
        "connections": connections,
    }


def main():
    drive = sys.argv[1] if len(sys.argv) > 1 else "knowledge-vault"

    # Get drive info and research folder
    tree_data, err = run_sw("docs", "tree", drive, "--format", "json")
    if not tree_data or isinstance(tree_data, str):
        print(f"Error reading drive: {err}")
        sys.exit(1)

    research_folder = None
    for node in tree_data.get("nodes", []):
        if node.get("name") == "research" and node.get("kind") == "folder":
            if not node.get("parentFolder"):
                research_folder = node["id"]
                break

    if not research_folder:
        print("Error: /research/ folder not found in drive")
        sys.exit(1)

    drive_id = tree_data["id"]
    print(f"Drive: {tree_data['name']} ({drive_id})")
    print(f"Research folder: {research_folder}")

    # Check existing claims
    existing = sum(
        1
        for n in tree_data.get("nodes", [])
        if n.get("documentType") == "bai/research-claim"
    )
    if existing >= 200:
        print(f"Methodology already imported ({existing} claims). Skipping.")
        sys.exit(0)

    # Build existing title set for skip-if-exists
    existing_titles = {
        n["name"]
        for n in tree_data.get("nodes", [])
        if n.get("documentType") == "bai/research-claim"
    }
    existing_id_map = {
        n["name"]: n["id"]
        for n in tree_data.get("nodes", [])
        if n.get("documentType") == "bai/research-claim"
    }

    # Parse all methodology files
    all_files = sorted([f for f in os.listdir(METHODOLOGY_DIR) if f.endswith(".md")])

    print(f"\nParsing {len(all_files)} methodology files...")
    claims = {}
    for filename in all_files:
        filepath = os.path.join(METHODOLOGY_DIR, filename)
        parsed = parse_methodology_file(filepath)
        if parsed:
            claims[parsed["title"]] = parsed

    print(f"Parsed {len(claims)} claims")

    # PASS 1: Create all claim documents (skips existing)
    print(f"\n{'='*60}")
    print(f"PASS 1: Creating research claim documents ({len(existing_titles)} already exist)")
    print(f"{'='*60}")

    title_to_id = dict(existing_id_map)  # Seed with existing docs
    created = 0
    skipped = 0
    failed = 0

    for i, (title, claim) in enumerate(claims.items()):
        # Skip if already exists in drive
        if title in existing_titles:
            skipped += 1
            if (i + 1) % 50 == 0:
                print(f"  [{i+1}/{len(claims)}] {created} created, {skipped} skipped, {failed} failed")
            continue

        # Create the document
        data, err = run_sw(
            "docs",
            "create",
            "--type",
            "bai/research-claim",
            "--name",
            title,  # Full title — truncation breaks wiki link resolution in pass 2
            "--drive",
            drive,
            "--parent-folder",
            research_folder,
            "--format",
            "json",
        )

        if data and isinstance(data, dict) and data.get("id"):
            doc_id = data["id"]
            title_to_id[title] = doc_id

            # Populate with CREATE_CLAIM
            claim_input = json.dumps(
                {
                    "title": claim["title"],
                    "description": claim["description"][:500] if claim["description"] else "Research claim",
                    "content": claim["content"],
                    "kind": claim["kind"],
                    "methodology": claim["methodology"],
                    "sources": claim["sources"],
                    "topics": claim["topics"],
                }
            )

            _, err2 = run_sw(
                "docs",
                "mutate",
                doc_id,
                "--op",
                "createClaim",
                "--input",
                claim_input,
                "--format",
                "json",
            )
            if err2 and "Error" in err2:
                print(f"  [{i+1}/{len(claims)}] {title[:50]}... CREATE_CLAIM failed: {err2[:80]}")
                failed += 1
            else:
                created += 1
                if (i + 1) % 10 == 0:
                    print(f"  [{i+1}/{len(claims)}] {created} created, {skipped} skipped, {failed} failed")
        else:
            print(f"  [{i+1}/{len(claims)}] FAILED to create: {title[:50]}... {err}")
            failed += 1

        # Delay to avoid overwhelming the reactor (longer for remote)
        time.sleep(0.3)

    print(f"\nPass 1 complete: {created} created, {failed} failed")
    print(f"Title→ID map: {len(title_to_id)} entries")

    # PASS 2: Resolve connections
    print(f"\n{'='*60}")
    print(f"PASS 2: Resolving connections")
    print(f"{'='*60}")

    resolved = 0
    unresolved = 0
    conn_errors = 0

    for i, (title, claim) in enumerate(claims.items()):
        if title not in title_to_id:
            continue
        source_id = title_to_id[title]

        for j, conn in enumerate(claim["connections"]):
            target_title = conn["target"]
            if target_title in title_to_id:
                target_id = title_to_id[target_title]
                # Use a deterministic hash of source+target+index for globally unique IDs
                conn_hash = hashlib.md5(f"{source_id}:{target_id}:{j}".encode()).hexdigest()[:16]
                conn_input = json.dumps(
                    {
                        "id": f"conn-{conn_hash}",
                        "targetRef": target_id,
                        "contextPhrase": conn["context"][:200],
                    }
                )
                _, err = run_sw(
                    "docs",
                    "mutate",
                    source_id,
                    "--op",
                    "addResearchConnection",
                    "--input",
                    conn_input,
                    "--format",
                    "json",
                )
                if err and "Error" in err:
                    conn_errors += 1
                else:
                    resolved += 1
            else:
                unresolved += 1

        if (i + 1) % 20 == 0:
            print(
                f"  [{i+1}/{len(claims)}] {resolved} resolved, {unresolved} unresolved, {conn_errors} errors"
            )

        # Delay to avoid overwhelming the reactor (longer for remote)
        time.sleep(0.2)

    print(f"\nPass 2 complete: {resolved} resolved, {unresolved} unresolved, {conn_errors} errors")

    print(f"\n{'='*60}")
    print(f"METHODOLOGY IMPORT COMPLETE")
    print(f"{'='*60}")
    print(f"Claims created: {created}")
    print(f"Connections resolved: {resolved}")
    print(f"Connections unresolved: {unresolved}")
    print(f"Errors: {failed + conn_errors}")


if __name__ == "__main__":
    main()

File: scripts/import-methodology.mjs
#!/usr/bin/env node
/**
 * Import Ars Contexta methodology (249 research claims) into a Knowledge Vault drive
 * via the Switchboard GraphQL API.
 *
 * Three-pass approach to avoid drive corruption:
 *   Pass 1: Create all documents standalone + populate with CREATE_CLAIM
 *   Pass 1b: Batch ADD_FILE actions on the drive (groups of 10, with delays)
 *   Pass 2: Resolve cross-claim connections
 *
 * Usage:
 *   node import-methodology.mjs --drive-id <UUID> --research-folder-id <UUID> [--endpoint <URL>] [--dry-run] [--limit N] [--batch-size N]
 */

import fs from "fs";
import path from "path";
import { fileURLToPath } from "url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const METHODOLOGY_DIR = path.resolve(__dirname, "..", "data", "methodology");

function getArg(flag, fallback) {
  const idx = process.argv.indexOf(flag);
  return idx !== -1 ? process.argv[idx + 1] : fallback;
}

const ENDPOINT = getArg("--endpoint", "https://switchboard-dev.powerhouse.xyz/graphql/r");
const DRIVE_ID = getArg("--drive-id", null);
const RESEARCH_FOLDER_ID = getArg("--research-folder-id", null);
const DRY_RUN = process.argv.includes("--dry-run");
const LIMIT = parseInt(getArg("--limit", "Infinity"), 10) || Infinity;
const BATCH_SIZE = parseInt(getArg("--batch-size", "10"), 10);

if (!DRIVE_ID || !RESEARCH_FOLDER_ID) {
  console.error(
    "Usage: node import-methodology.mjs --drive-id <UUID> --research-folder-id <UUID> [--endpoint <URL>] [--dry-run] [--limit N] [--batch-size N]",
  );
  process.exit(1);
}

// --- Helpers ---

async function gql(query, variables) {
  const res = await fetch(ENDPOINT, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ query, variables }),
  });
  const json = await res.json();
  if (json.errors) {
    throw new Error(JSON.stringify(json.errors, null, 2));
  }
  return json.data;
}

function parseFrontmatter(content) {
  const match = content.match(/^---\n([\s\S]*?)\n---\n([\s\S]*)$/);
  if (!match) return { meta: {}, body: content };

  const rawMeta = match[1];
  const body = match[2].trim();
  const meta = {};

  for (const line of rawMeta.split("\n")) {
    const m = line.match(/^(\w+):\s*(.+)$/);
    if (m) {
      const [, key, val] = m;
      if (val.startsWith("[")) {
        meta[key] = val
          .replace(/[\[\]]/g, "")
          .split(",")
          .map((s) => s.trim().replace(/^["']|["']$/g, "").replace(/^\[\[|\]\]$/g, ""))
          .filter(Boolean);
      } else {
        meta[key] = val.replace(/^\[\[|\]\]$/g, "").replace(/^["']|["']$/g, "");
      }
    }
  }

  return { meta, body };
}

function extractRelevantNotes(body) {
  const section = body.match(/Relevant Notes:\n([\s\S]*?)(?:\n\nTopics:|\n---|\n$|$)/);
  if (!section) return [];
  const links = [];
  for (const line of section[1].split("\n")) {
    const m = line.match(/- \[\[(.+?)\]\]\s*[—–-]\s*(.+)/);
    if (m) links.push({ title: m[1].trim(), contextPhrase: m[2].trim() });
  }
  return links;
}

function extractContentBody(body) {
  let content = body.replace(/^#\s+.+\n\n?/, "");
  const rnIdx = content.indexOf("\nRelevant Notes:");
  if (rnIdx !== -1) content = content.substring(0, rnIdx);
  const sepIdx = content.indexOf("\n---\n---");
  if (sepIdx !== -1) content = content.substring(0, sepIdx);
  return content.trim();
}

function delay(ms) {
  return new Promise((r) => setTimeout(r, ms));
}

function now() {
  return new Date().toISOString();
}

// --- Main ---

async function main() {
  const files = fs
    .readdirSync(METHODOLOGY_DIR)
    .filter((f) => f.endsWith(".md"))
    .slice(0, LIMIT);

  console.log(`Found ${files.length} methodology files`);
  console.log(`Endpoint: ${ENDPOINT}`);
  console.log(`Drive: ${DRIVE_ID}`);
  console.log(`Research folder: ${RESEARCH_FOLDER_ID}`);
  console.log(`Batch size for ADD_FILE: ${BATCH_SIZE}`);
  if (DRY_RUN) console.log("DRY RUN — no documents will be created");
  console.log();

  // ========================================
  // Pass 1: Create documents + populate
  // (NO drive mutations here — standalone docs only)
  // ========================================
  console.log("=== Pass 1: Create and populate documents ===");
  const titleToId = new Map();
  const titleToConnections = new Map();
  const createdDocs = []; // { docId, title } for Pass 1b
  let created = 0;
  let failed = 0;

  for (const file of files) {
    const title = file.replace(/\.md$/, "");
    const raw = fs.readFileSync(path.join(METHODOLOGY_DIR, file), "utf-8");
    const { meta, body } = parseFrontmatter(raw);
    const content = extractContentBody(body);
    const connections = extractRelevantNotes(body);
    titleToConnections.set(title, connections);

    const description = meta.description || "";
    const kind = meta.kind || "research";
    const methodology = Array.isArray(meta.methodology) ? meta.methodology : meta.methodology ? [meta.methodology] : [];
    const sources = Array.isArray(meta.source) ? meta.source : meta.source ? [meta.source] : [];
    const topics = Array.isArray(meta.topics) ? meta.topics : meta.topics ? [meta.topics] : [];

    if (DRY_RUN) {
      console.log(`  [dry] ${title} (${connections.length} connections, ${topics.length} topics)`);
      titleToId.set(title, "dry-run-id");
      createdDocs.push({ docId: "dry-run-id", title });
      continue;
    }

    try {
      // Create empty document (standalone — no parent, no drive mutation)
      const createData = await gql(
        `mutation($docType: String!) {
          createEmptyDocument(documentType: $docType) { id }
        }`,
        { docType: "bai/research-claim" },
      );
      const docId = createData.createEmptyDocument.id;

      // Populate claim immediately
      await gql(
        `mutation($id: String!, $actions: [JSONObject!]!) {
          mutateDocument(documentIdentifier: $id, actions: $actions) { id }
        }`,
        {
          id: docId,
          actions: [
            {
              type: "CREATE_CLAIM",
              input: { title, description, content, kind, methodology, sources, topics },
              scope: "global",
              timestampUtcMs: now(),
            },
          ],
        },
      );

      titleToId.set(title, docId);
      createdDocs.push({ docId, title });
      created++;
      if (created % 10 === 0) console.log(`  Created ${created}/${files.length}...`);

      // Small delay between creates
      await delay(50);
    } catch (err) {
      console.error(`  FAIL: ${title}: ${err.message.substring(0, 200)}`);
      failed++;
    }
  }

  console.log(`\nPass 1 complete: ${created} created, ${failed} failed\n`);

  // ========================================
  // Pass 1b: Batch ADD_FILE on the drive
  // (groups of BATCH_SIZE with 500ms between batches)
  // ========================================
  console.log(`=== Pass 1b: Add files to drive in batches of ${BATCH_SIZE} ===`);

  if (!DRY_RUN) {
    let addedToDrive = 0;
    for (let i = 0; i < createdDocs.length; i += BATCH_SIZE) {
      const batch = createdDocs.slice(i, i + BATCH_SIZE);
      const actions = batch.map((doc) => ({
        type: "ADD_FILE",
        input: {
          id: doc.docId,
          name: doc.title,
          documentType: "bai/research-claim",
          parentFolder: RESEARCH_FOLDER_ID,
        },
        scope: "global",
        timestampUtcMs: now(),
      }));

      try {
        await gql(
          `mutation($id: String!, $actions: [JSONObject!]!) {
            mutateDocument(documentIdentifier: $id, actions: $actions) { id }
          }`,
          { id: DRIVE_ID, actions },
        );
        addedToDrive += batch.length;
        console.log(`  Added batch ${Math.floor(i / BATCH_SIZE) + 1}: ${addedToDrive}/${createdDocs.length} files in drive`);
      } catch (err) {
        console.error(`  FAIL batch at index ${i}: ${err.message.substring(0, 300)}`);
      }

      // Wait between drive mutations to avoid corruption
      await delay(500);
    }
    console.log(`\nPass 1b complete: ${addedToDrive} files added to drive\n`);
  } else {
    console.log(`  [dry] Would add ${createdDocs.length} files in ${Math.ceil(createdDocs.length / BATCH_SIZE)} batches\n`);
  }

  // ========================================
  // Pass 2: Resolve connections
  // ========================================
  console.log("=== Pass 2: Resolve connections ===");

  if (DRY_RUN) {
    let totalConns = 0;
    for (const [, conns] of titleToConnections) totalConns += conns.length;
    console.log(`  [dry] Would create ${totalConns} connections`);
    return;
  }

  let resolved = 0;
  let unresolved = 0;

  for (const [title, connections] of titleToConnections) {
    const sourceId = titleToId.get(title);
    if (!sourceId) continue;

    const actions = [];
    for (const conn of connections) {
      const targetId = titleToId.get(conn.title);
      if (targetId) {
        actions.push({
          type: "ADD_RESEARCH_CONNECTION",
          input: {
            id: crypto.randomUUID(),
            targetRef: targetId,
            contextPhrase: conn.contextPhrase,
          },
          scope: "global",
          timestampUtcMs: now(),
        });
        resolved++;
      } else {
        unresolved++;
      }
    }

    if (actions.length > 0) {
      try {
        await gql(
          `mutation($id: String!, $actions: [JSONObject!]!) {
            mutateDocument(documentIdentifier: $id, actions: $actions) { id }
          }`,
          { id: sourceId, actions },
        );
      } catch (err) {
        console.error(`  FAIL connections for ${title}: ${err.message.substring(0, 200)}`);
      }
      await delay(50);
    }
  }

  console.log(`\nPass 2 complete: ${resolved} connections resolved, ${unresolved} unresolved`);
  console.log(`\n=== METHODOLOGY IMPORT COMPLETE ===`);
  console.log(`Claims: ${created}`);
  console.log(`Connections: ${resolved} resolved, ${unresolved} unresolved`);
  console.log(`Drive mutations: ${Math.ceil(createdDocs.length / BATCH_SIZE)} batches (instead of ${createdDocs.length} individual)`);
}

main().catch((err) => {
  console.error("Fatal:", err);
  process.exit(1);
});
